Initial airflow airbyte tutorial (#2191)
airbytehq/airbyte: docs/tutorials/airflow-trigger-airbyte-job.md

---
description: Create a DAG in Airflow to trigger Airbyte sync job
---

# Airflow Integration 

> Airbyte allows to trigger the syncronization job using Operators in Airflow. This tutorials explain how to configure your DAG in Airflow to do this. 

Reviewer: ```suggestion
Airbyte allows you to trigger synchronization jobs using Airflow Operators. This tutorial explains how to configure your Airflow DAG to do so. 
```

## 1. Set up the tools

> First of all, make sure you have Docker and Docker Compose installed. 

Reviewer: ```suggestion
First of all, make sure you have Docker installed. (We'll be using the `docker-compose` command, so please make sure your install contains `docker-compose`.)
```

### **Start Airbyte**

> For this tutorial you need have a connection created. If you are starting with Airbyte and want to test the integration, we suggest to you follow the [Getting Started tutorial](../getting-started.md) and use that connection.

Reviewer: ```suggestion
If this is your first time using Airbyte, we suggest first go through our [basic tutorial](../getting-started.md). This tutorial will use the connection set up in the basic tutorial.

If you are an experienced user, carry on!
```

> The recommendation is to set up the **sync frequency** of your Connection to **manual** because Airflow will become responsible to trigger this process.

Reviewer: @marcosmarxm what happens if we don't set sync to manual? Will this cause problems?

Author: Airbyte will trigger automatic that sync usually. It's not a problem but can generate some confusion.

Think that you set the **sync frequency** to 1h in Airybte, but your Airflow job is `@daily`
And you have a pipeline e.g `airbyte_sync_job >> aggregate_data >> send_report_email` in Airflow.
When you start your daily job in Airflow it'd already triggered 23x in Airbyte.

Could be a problem with you have a dashboard where you use the raw data collected in the `airbyte_sync_job` and also the result from `aggregate_data`. Why?
In the dashboard, a graph using `airbyte_sync_job` has 1h delay, but another graph using the `aggregate_data` has D-1 delay because is only updated by Airflow.

### **Start Airflow**

> To setup Airflow we recommend you to follow this [Quick start Airflow](https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html) if you don't have any instances yet.

Reviewer: ```suggestion
If you don't have an Airflow instance, we recommend following this [Quick start Airflow](https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html) to set one up.
```

## 2. Create a DAG in Airflow to trigger your Airbyte job

> You can access Airflow at [http://localhost:8080/](http://localhost:8080/).

Reviewer: ```suggestion
```

> ### Create Airbyte connection in Airflow

Reviewer: ```suggestion
### Create an Airbyte connection in Airflow
```

> After Airflow starts, access the connections page to create our connection with Airbyte.

Reviewer: ```suggestion
Once Airflow starts, navigate to Airflow's `Connections` page. You can access Airflow at [http://localhost:8080/](http://localhost:8080/).
```

![](../.gitbook/assets/airflow_create_connection.png)

> Insert data about our connection. We are going to use the Airbyte API to execute our actions. For this reason we will create an HTTP connection.

Reviewer: ```suggestion
Airflow will use the Airbyte API to execute our actions. For this reason, we will create a HTTP connection. The typical Airbyte set up is hosted at `localhost:8001`. Configure Airflow's HTTP connection accordingly - we've provide a screenshot example.
```

![](../.gitbook/assets/airflow_edit_connection.png)

> After inserting the information click on Save.

Reviewer: ```suggestion
After inserting the information, click Save.
```

> ### Getting information about your Airbyte connection job

Reviewer: ```suggestion
### Retrieving the Airbyte Connection ID
```

> You can access Airbyte at `localhost:8000`. 

Reviewer: ```suggestion
We'll need the Airbyte connection id so our Airflow DAG knows which Airbyte connection to trigger.
```

![](../.gitbook/assets/airflow_airbyte_connection.png)

> That code in the url is the `connectionId` used by Airflow to trigger the job. In our example: `1e3b5a72-7bfd-4808-a13c-204505490110`

Reviewer: ```suggestion
This id can be seen in the URL on the connection page in the Airbyte UI. To get to the connection page, first click on a source. This should bring up a Connections table with a list of configured connections. Click on the connection in question.

This id is `1e3b5a72-7bfd-4808-a13c-204505490110` in our example. 
```

> ### Create a simple DAG in Airflow to run a sync job

Reviewer: ```suggestion
### Create an Airflow DAG to run an Airbyte sync job
```

> Create a file inside the `/dags` directory called `dag_airbyte_example.py`

Reviewer: ```suggestion
Place the following file inside the `/dags` directory. Name this file `dag_airbyte_example.py`.
```

```python
from airflow import DAG
from airflow.utils.dates import days_ago
from airflow.providers.airbyte.operator import AirbyteTriggerSyncOperator

with DAG(dag_id='trigger_airbyte_job_example',
         default_args={'owner': 'airflow'},
         schedule_interval='@daily',
         start_date=days_ago(1)
    ) as dag:

    money_to_json = AirbyteTriggerSyncOperator(
        task_id='airbyte_money_json_example',
        airbyte_conn_id='airbyte_conn_example',
        connection_id='1e3b5a72-7bfd-4808-a13c-204505490110',
        asynchronous=False,
        timeout=360,
        wait_seconds=5
    )
```

> About the parameters used in the Operator:

Reviewer: ```suggestion
The Airbyte Airflow Operator accepts the following parameters:
```

> - `airbyte_conn_id`: the HTTP connection configured above.

Reviewer: ```suggestion
- `airbyte_conn_id`: name of the Airflow HTTP Connection configured to hit Airbyte.
```

> - `connection_id`: The uuid code for the connection job. You can get this in the url path from Airbyte UI.

Reviewer: ```suggestion
- `connection_id`: The ID of the Airbyte Connection to be triggered by Airflow.
```

> - `asynchronous`: this is specific to Airflow DAG execution. Default is `false`. This is useful for submitting long running jobs and         waiting on them asynchronously using the **AirbyteJobSensor**. If you have a very busy Airflow instance this could be handful because after submiting the job you use the sensor to check, and a sensor doesn't occupy a worker slot in Airflow. Later we show how to configure a DAG using this configuration.

Reviewer: ```suggestion
- `asynchronous`: Determines how the Airbyte Operator runs. When true, Airflow will monitor the Airbyte Job via the **AirbyteJobSensor**. Since sensors do not occupy an Airflow worker slot, this is helps reduce Airflow load. Default is `false.
```

> - `timeout`: maximum time to wait to finish the job to complete. This argument is only used when `asynchronous=False`.

Reviewer: ```suggestion
- `timeout`: Maximum time Airflow will wait for the Airbyte job to complete. Only valid when `asynchronous=False`.
```

> - `wait_seconds`: The amount of time to wait between checks. This argument is only used when `asynchronous=False`.

Reviewer: ```suggestion
- `wait_seconds`: The amount of time to wait between checks. Only valid when `asynchronous=False`.
```

> After creating the file our DAG will show up in Airflow UI. Click on the button shown below to actived our DAG. Soon our DAG will be trigger.

Reviewer: ```suggestion
Our DAG will show up in Airflow UI shortly after we place our DAG file and can be activated by clicking the button shown below. Airflow will trigger the DAG shortly after.
```

![](../.gitbook/assets/airflow_unpause_dag.png)

> You can check the job got start sync in Airbyte UI.

Reviewer: ```suggestion
Check to see if the job started syncing in the Airbyte UI's Sync History tab!
```

![](../.gitbook/assets/airflow_airbyte_trigger_job.png)

> ### How to create your DAG using the async parameter

Reviewer: ```suggestion
### Using the `asynchronous` parameter
```

> As commented before, if your Airflow instance is limited or already had alot of DAGs running you can you this option to trigger a more performatic DAG.

Reviewer: ```suggestion
If your Airflow instance has limited resources and/or is under load, setting the `asynchronous=True` can help.
```

```python
from airflow import DAG
from airflow.utils.dates import days_ago
from airflow.providers.airbyte.operator import AirbyteTriggerSyncOperator

with DAG(dag_id='airbyte_trigger_job_example_async',
         default_args={'owner': 'airflow'},
         schedule_interval='@daily',
         start_date=days_ago(1)
    ) as dag:

    async_money_to_json = AirbyteTriggerSyncOperator(
        task_id='airbyte_async_money_json_example',
        airbyte_conn_id='airbyte_conn_example',
        connection_id='1e3b5a72-7bfd-4808-a13c-204505490110',
        asynchronous=True,
    )

    airbyte_sensor = AirbyteJobSensor(
        task_id='airbyte_sensor_money_json_example',
        airbyte_conn_id='airbyte_conn_example',
        airbyte_job_id=async_money_to_json.output
    )

    async_money_to_json >> airbyte_sensor
```

## That's it!

> This is just the beginning of using Airbyte. We support a large collection of sources and destination. You can even contribute your own.

Reviewer: ```suggestion
This is just the beginning of using Airbyte. We support a large collection of sources and destinations. You can even contribute your own.
```
